feat: Introduce example of using VS Code with codetogether (#26)
eclipse-che/blog: _posts/2022-04-01-introducing-pair-programming.adoc

---
title: Pair programming using VS Code extension
layout: post
author: Florent Benoit
description: >-
  How to do pair programming using a VS Code extension.
categories: []
keywords: ['VSCode', 'codetogether', 'pair programming']
slug: /@florent.benoit/pair-programming-using-vscode-editor
---

One of the most wanted feature on Eclipse Che is to be able to do pair programming.

This blog post describes how it is possible to achieve that by using a third-party extension, link:https://open-vsx.org/extension/genuitecllc/codetogether[CodeTogether].

There are different pricing plans available for this extension, but there is one free item when connecting to a maximum of 60 minutes and 3 shares which may be enough for most of the case.

This is only possible using Eclipse Che starting with 7.46+ and using Visual Studio Code editor, not Che-Theia.

== Quick start from an example

. Go the dashboard page of Eclipse Che. For example URL will be following this format https://eclipse-che-my-instance.com/dashboard/#/

> . Replace the last part of the URL containing `/dashboard/#` by `#https://github.com/benoitf/codetogether-che-petclinic`. URL should look like `https://eclipse-che-my-instance.com#https://github.com/benoitf/codetogether-che-petclinic``
> 
> . Press enter key to open this URL in the browser.
> 

Reviewer: ```suggestion
. Replace the last part of the URL containing `/dashboard/#` with `#https://github.com/benoitf/codetogether-che-petclinic` and press enter.

```

. Workspace will start and it will open automatically Visual Studio Code using a sample project based on `Petclinic` application.

. As link:https://open-vsx.org/extension/genuitecllc/codetogether[CodeTogether] is a recommended extension, accept the prompt asking to install this extension.

. CodeTogether will ask to access Visual Studio Code, click on `No`

. A new prompt will ask to redirect port 1593: Click on `No`

. Bring the command palette using for example `F1` key and select `CodeTogether: Start Hosting Session`

> . Keep default on all properties and click on the `start` button

Reviewer: ```suggestion
. Keep the defaults for all properties and click on the `start` button
```

. Click "No" on "CodeTogether requires access to Visual Studio Code Proposed APIs"

> . `Pair programming` link is then available in the clipboard.

Reviewer: ```suggestion
. A `Pair programming` link is then available in the clipboard.
```

> . Now, share this link with your team members. They will be able to join your workspace.

Reviewer: ```suggestion
. You can share this link with your team members. They will be able to join your workspace.
```

> . It is possible to join from different IDEs or from a running Eclipse Che workspace (like the one we are using). But click on joining a session rather than starting a new session: 'CodeTogether: Join Session' and paste the link in the input field.

Reviewer: ```suggestion
. It is possible to use this link to join from different IDEs or from a running Eclipse Che workspace (like the one we are using). To do that select 'CodeTogether: Join Session' rather than starting a new session.
```

== Enabling pair programming settings on your own example

> In the previous section, a pre-configured workspace was used but it is also possible to use your own example

Reviewer: ```suggestion
In the previous section, a pre-configured repository was used but it is also possible to use your own git repository.
```

This feature works only with Visual Studio Code IDE (not for example with Che-Theia).

> === Use Visual Studio Code as IDE inside Eclipse Che

Reviewer: ```suggestion
=== Use Visual Studio Code as the IDE of an Eclipse Che workspace
```

> To use Visual Studio Code IDE when starting a workspace there are two ways: 

Reviewer: ```suggestion
There are 2 ways to use Visual Studio Code IDE in a Che workspace: 
```

> * Adds `?che-editor=che-incubator/che-code/insiders` at the end of your existing workspace URL

Reviewer: ```suggestion
* Add `?che-editor=che-incubator/che-code/insiders` at the end of the workspace URL
```

> * Adds a `.che/che-editor.yaml` file with the following content:

Reviewer: ```suggestion
* Add a `.che/che-editor.yaml` file with the following content:
```

+
[source,yaml,subs="+quotes"]
----

> sid: che-incubator/che-code/insiders

Reviewer: ```suggestion
id: che-incubator/che-code/insiders
```

----

> === Enable CodeTogether as recommended extension

Reviewer: ```suggestion
=== Recommend the installation of CodeTogether
```

> Adds CodeTogether as being a recommended extension.

Reviewer: ```suggestion
Include CodeTogether among the recommended extension.
```

> . Create `.vscode/extensions.json` file with the following content:

Reviewer: ```suggestion
. Create the file `.vscode/extensions.json` with the following content:
```

+
[source,json,subs="+quotes"]
----
  {
    "recommendations": [
      "genuitecllc.codetogether",
    ]
  }
----

> Then CodeTogether extension is now a recommended extension and on workspace startup it will prompt the user to install it.

Reviewer: ```suggestion
That will trigger a prompt to install CodeTogether at workspace startup.
```

> === Adds a devfile

Reviewer: ```suggestion
=== Add a Devfile
```

> Probably you don't have any devfile in your repository so you should add a very simple one using the Eclipse Che universal developer image (most of common tools already installed)

Reviewer: ```suggestion
If you don't have a Devfile at the root of your repository you should add one with a https://github.com/devfile/developer-images/[Universal Developer Image] component (it includes the most popular development tools and languages).
```

. Create a `.devfile.yaml` file with the following content:
+
[source,yaml,subs="+quotes"]
----
schemaVersion: 2.1.0
metadata:
  name: my-workspace
components:
  - name: tools
    container:
      image: quay.io/devfile/universal-developer-image:ubi8-0e189d9
      memoryLimit: 3Gi

----

. Do not forget to update the `name` attribute of the devfile to match your project.

=== Share your workspace

> Once you have committed these changes on a repository, you can share your workspace with your team members by sharing the URL to your repository.

Reviewer: ```suggestion
Once you have committed these changes on a repository, you can share your workspace URL with your team.
```

> URL will be like this: `https://eclipse-che-my-instance.com#https://github.com/<your-repository>`

Reviewer: ```suggestion
The URL looks like this: `https://eclipse-che-my-instance.com#https://github.com/<your-repository>`
```
